### PR TITLE
Feature tmc#144 float literal as const token

### PR DESCRIPTION
**Pull request by gawkowsk**
float const literals are supported in format digit{digit}'.'{digit}[(e|E)[(+|-)]digit{digit}]
comparison and arithmetic operators handled correctly, testing for tokenization of float const literals and operators are added

### CODE REVIEW

File: .vscode/launch.json [hunk 1]
```diff
@@ -0,0 +1,36 @@
+{
```
Comment (Lurkerpas) [inline .vscode/launch.json:1]: I think that VSC settings should be private for a user, and should not be pushed to the repository.

Comment (Lurkerpas) [inline .vscode/launch.json]: I think that VSC settings should be private for a user, and should not be pushed to the repository.

File: Src/mesg.c [hunk 2]
```diff
@@ -763,7 +794,7 @@ doq(Symbol *s, int n, RunList *r)
 			for (j = 0; j < q->nflds; j++)
 			{	if (j > 0) printf(",");
 				sr_mesg(stdout,
-					getInt(q->contents[k*q->nflds+j]),
+					getInt(q->contents[k*q->nflds+j]),	// TODO PG - handle float in queues?
```
Comment (Lurkerpas) [inline Src/mesg.c:797]: Yes, floats in queues are a must-have feature.

File: Src/pangen2.c [hunk 3]
```diff
@@ -1947,7 +1947,8 @@ generic_case:	fprintf(fd_tt, "\ttrans[%d][%d]\t= ", Pid_nr, e->seqno);
 		goto non_generic;
 #endif
 	case 'c':
-		if (e->n->lft->ntyp == CONST
+		if (e->n->lft->ntyp == CONST 
+		&& e->n->lft->constValKind == VALUE_INT		// TODO PG - verify what to do if float?
```
Comment (Lurkerpas) [inline Src/pangen2.c:1951]: It would be easier to answer this question, if the purpose of this function and this particular case was known. Should we discuss this?

File: Src/pangen2.c [hunk 4]
```diff
@@ -2066,6 +2067,7 @@ in_settr--;
 
 				if (g->n->ntyp == 'c'
 				&&  g->n->lft->ntyp == CONST
+				&& 	g->n->lft->constValKind == VALUE_INT		// TODO PG - verify what to do if float
```
Comment (Lurkerpas) [inline Src/pangen2.c]: Indent. Also, see the previous comment.

File: Src/pangen3.c [hunk 5]
```diff
@@ -296,7 +296,10 @@ comwork(FILE *fd, Lextok *now, int m)
 			&&  now->sym->mtype_name)
 			{	s = now->sym->mtype_name->name;
 			}
-			sr_mesg(fd, now->val, now->ismtyp, s);
+			if (now->constValKind == VALUE_FLOAT)
+				sr_mesg_f(fd, getFloatTokenValue(now), now->ismtyp, s);
+			else
+				sr_mesg(fd, now->val, now->ismtyp, s);	// TODO PG - inside there is comparison of now->val to 1 and printf as int...
```
Comment (Lurkerpas) [inline Src/pangen3.c:302]: I don't understand the comment in the context of the current version of the code.

File: Src/pangen7.c [hunk 6]
```diff
@@ -193,10 +193,12 @@ complete_transition(Succ_List *sl, Guard *g)
 	printf("	:: ");
 	for (w = g; w; w = w->nxt)
 	{	if (w->t->ntyp == CONST
+		&&  w->t->constValKind == VALUE_INT	// TODO PG - verify if it could be float...
 		&&  w->t->val == 1)
 		{	continue;
 		} else if (w->t->ntyp == 'c'
 		&&  w->t->lft->ntyp == CONST
+		&& 	w->t->lft->constValKind == VALUE_INT	// TODO PG - verify if it could be float... and then what?
```
Comment (Lurkerpas) [inline Src/pangen7.c]: Ident. Also, this is something that we could discuss.

File: Src/pangen7.c [hunk 7]
```diff
@@ -674,7 +676,8 @@ create_transition(OneState *s, SQueue *it)
 		{	wrap_text("", t, " ");
 		}
 		if (t->ntyp == 'c'
-		&&  t->lft->ntyp == CONST)
+		&&  t->lft->ntyp == CONST
+		&&	t->lft->constValKind == VALUE_INT)	// TODO PG - could be float? and then what?
```
Comment (Lurkerpas) [inline Src/pangen7.c]: Indent.

File: Src/sched.c [hunk 8]
```diff
@@ -19,7 +19,7 @@ extern int	lineno, nr_errs, dumptab, xspin, jumpsteps, columns;
 extern int	u_sync, Elcnt, interactive, TstOnly, cutoff;
 extern short	has_enabled, has_priority, has_code, replay;
 extern int	limited_vis, product, nclaims, old_priority_rules;
-extern int	old_scope_rules, scope_seq[128], scope_level, has_stdin;
+extern int	old_scope_rules, scope_seq[256], scope_level, has_stdin;
```
Comment (Lurkerpas) [inline Src/sched.c:22]: If we are changing such magic number anyways, maybe let's name them e.g. via define?

File: Src/spin.h [hunk 9]
```diff
@@ -27,6 +27,7 @@ typedef struct Lextok {
 	unsigned short	ntyp;	/* node type */
 	short	ismtyp;		/* CONST derived from MTYP */
 	int	val;		/* value attribute */
+	ValueKind constValKind;
```
Comment (Lurkerpas) [inline Src/spin.h]: Doc, for consistency.

File: Src/spin.h [hunk 10]
```diff
@@ -316,8 +317,9 @@ int	eval(Lextok *);
 int	find_lab(Symbol *, Symbol *, int);
 int	find_maxel(Symbol *);
 int	full_name(FILE *, Lextok *, Symbol *, int);
-Value getlocal(Lextok *);
-Value getval(Lextok *);
+Value getlocal(Lextok *);	// TODO PG - actually it takes the value from the symbol in the token... rename
```
Comment (Lurkerpas) [inline Src/spin.h]: Indent.

File: Src/spin.y [hunk 11]
```diff
@@ -488,11 +489,17 @@ ivar    : vardcl           	{ $$ = $1;
 				  nochan_manip($1, $3, 0);
 				  no_internals($1);
 				  if (!initialization_ok)
-				  {	Lextok *zx = nn(ZN, NAME, ZN, ZN);
-					zx->sym = $1->sym;
-					add_seq(nn(zx, ASGN, zx, $3));
-					$1->sym->ini = 0;	/* Patrick Trentlin */
-				  }
+				  {	if ($1->sym->isarray)
+					{	fprintf(stderr, "warning: %s:%d initialization of %s[] ",
```
Comment (Lurkerpas) [inline Src/spin.y:493]: I don't undestand this change. Could you please add a comment in code?

File: Src/spinlex.c [hunk 12]
```diff
@@ -51,9 +52,9 @@ extern int	implied_semis, ltl_mode, in_seq, par_cnt;
 
 short	has_stack = 0;
 int	lineno  = 1;
-int	scope_seq[128], scope_level = 0;
+int	scope_seq[256], scope_level = 0;
```
Comment (Lurkerpas) [inline Src/spinlex.c:55]: As stated earlier, if we are changing magic numbers, it would be nice to name them via defines. I see this is already done e.g., for MAX_TOKEN_TEXT_LENGTH

File: Src/spinlex.c [hunk 13]
```diff
@@ -1575,15 +1611,43 @@ lex(void)
...
 
 	if (isdigit_(c))
-	{	long int nr;
-		getword(c, isdigit_);
-		errno = 0;
-		nr = strtol(yytext, NULL, 10);
-		if (errno != 0)
-		{	fprintf(stderr, "spin: value out of range: '%s' read as '%d'\n",
-				yytext, (int) nr);
+	{	getword(c, isdigit_);
```
Comment (Lurkerpas) [inline Src/spinlex.c:1614]: Is this feasible to refactor this piece of code? This is an algorithm that we add, not just adapt, so I think we could try to follow good programming practices, if possible.
Can this code (or its parts) be extracted to reduce nesting and cyclomatic complexity?
Can we avoid reassining new values to existing variables - this is dangerous, especially as we have consecutive ifs without returns which rely on a value that is changed inside.

File: Src/spinlex.c [hunk 14]
```diff
@@ -1667,6 +1731,8 @@ lex(void)
 	{	switch (c) {
 		case '-': c = follow('>', IMPLIES,    '-'); break;
 		case '[': c = follow(']', ALWAYS,     '['); break;
+		case '/': c = follow('\\', AND, '/'); break;
+		case '\\': c = follow('/', OR, '\\'); break;
```
Comment (Lurkerpas) [inline Src/spinlex.c:1735]: What is the purpose of this change?

File: Src/structs.c [hunk 15]
```diff
@@ -559,7 +560,7 @@ dump_struct(Symbol *z, char *prefix, RunList *r)
 					}
 
 					sr_mesg(stdout, getInt(tl->sym->val[jx]),
-						tl->sym->type == MTYPE, s);
+						tl->sym->type == MTYPE, s);		// TODO PG - handle floats in queues? structs?
```
Comment (Lurkerpas) [inline Src/structs.c:563]: Yes, floats should be handled in structs and queues.

File: Src/value.h [hunk 16]
```diff
@@ -10,7 +10,7 @@
 #define SEEN_VALUE_H
 
 /* Types of primitives held in Value type*/
-typedef enum {VALUE_FLOAT, VALUE_INT} ValueKind;
+typedef enum {VALUE_INT, VALUE_FLOAT} ValueKind;
```
Comment (Lurkerpas) [inline Src/value.h:13]: I am curious about the reasons for this change. Does it mask some issues, e.g., when Kind is not initialized/initialized to 0?

File: Test/Float/FloatArithmetic.pml [hunk 17]
```diff
@@ -19,15 +19,15 @@ active proctype test() {
     float rdiv_lower;
     makeFloat(rdiv_lower, 74, 100);
     makeFloat(rdiv_upper, 76, 100);
-    a = 12;
```
Comment (Lurkerpas) [inline Test/Float/FloatArithmetic.pml:22]: I think we could remove makeFloat now, right?

File: Test/Float/FloatArithmeticOperators.pml [hunk 18]
```diff
@@ -0,0 +1,55 @@
+// This is needed because dot notation is not yet supported by the lexer
+hidden float tmpN;
+hidden float tmpD;
+
+inline makeFloat(f, nominator, denominator) {
```
Comment (Lurkerpas) [inline Test/Float/FloatArithmeticOperators.pml]: This is no longer needed, right?

Comment (Lurkerpas) [reply Test/Float/FloatArithmeticOperators.pml]: This includes the inline, hidden floats and the comment above.

File: Test/Float/FloatArithmeticOperators.pml [hunk 19]
```diff
@@ -0,0 +1,55 @@
...
+    assert(14.0 != 0);
+    assert(a < 15);
+    assert(a > c);
+    assert(a >= c);
+    assert(a <= 15)
+
+    // float var is properly compared to float
+    assert(a != b);
+    assert(b == b);
+    assert(a < 15.6);
```
Comment (Lurkerpas) [inline Test/Float/FloatArithmeticOperators.pml]: Could we check against 14.3?

File: Test/Float/FloatArithmeticOperators.pml [hunk 20]
```diff
@@ -0,0 +1,55 @@
...
+    assert(45.0 / 2 == 22.5);
+        // float can be devided by float
+    assert(45.0 / 2.0 == 22.5);
+
+    // int can be devided by int giving float
+    assert(45 / 2 == 22);
+    // int can be devided by float giving float
+    assert(45 / 2.0 == 22.5);
+
+    printf("Float arithmetic operators:\n a == %f, b == %f\n",
```
Comment (Lurkerpas) [inline Test/Float/FloatArithmeticOperators.pml]: I am not sure whether this text provides value.

Comment (Lurkerpas) [reply Test/Float/FloatArithmeticOperators.pml]: Please consider that the previous printf informed about both the values, and their source:
  printf("Float arithmetic: %f and %f == (+)%f (-)%f (*)%f (/)%f\n"....

File: Test/Float/FloatLiteralValues.pml [hunk 21]
```diff
@@ -0,0 +1,44 @@
...
+
+    x = 520e-1;
+    printf("x = %f\n", x);
+    assert(x - 52.0 < epsilon);
+
+    float a=3.0e+1 / 10;
+    float b=12.0;
+    float c;
+    c = a + b - 0.01;
+    printf("c variable has value of %f\n", c);
```
Comment (Lurkerpas) [inline Test/Float/FloatLiteralValues.pml]: I am not sure about the value of the prints in the current form. Please, either remove them, or provide additional information in them.

File: Test/Makefile [hunk 22]
```diff
@@ -439,6 +441,44 @@ benchmark_large_message:
...
+	${SPIN} ${FLOAT_MODEL_DIR}/FloatLiteralValues.pml > FloatLiteralValues.sim.output
+	grep -q "State-vector 40 byte, depth reached 2, errors: 0" ${TEST_DIR}/FloatLiteralValues.output
+	grep -q "3 states, stored" ${TEST_DIR}/FloatLiteralValues.output
+	grep -q "0 states, matched" ${TEST_DIR}/FloatLiteralValues.output
+	grep -q "3 transitions (= stored+matched)" ${TEST_DIR}/FloatLiteralValues.output
+	grep -q "0 atomic steps" ${TEST_DIR}/FloatLiteralValues.output
+	grep -q "x = 123.000000" ${TEST_DIR}/FloatLiteralValues.sim.output	
+	grep -q "y = 43.000000" ${TEST_DIR}/FloatLiteralValues.sim.output	
+	grep -q "z = 56.099998" ${TEST_DIR}/FloatLiteralValues.sim.output	
+	grep -q "x = 56.130001" ${TEST_DIR}/FloatLiteralValues.sim.output
```
Comment (Lurkerpas) [inline Test/Makefile]: I think that checking if X == A and then X == B, where A != B looks suspicious. I would prefer to modifiy the prints to uniquely indentify the checked value.

File: Src/spinlex.c (hunk 14 above)
Comment (gawkowsk) [reply Src/spinlex.c:1735]: It is not mine - this change is merged here from nimble repo before I've started any of my changes

File: Src/spinlex.c (hunk 12 above)
Comment (gawkowsk) [reply Src/spinlex.c:55]: Agree - but, it is not mine - this change is merged here from nimble repo before I've started any of my changes

File: Src/spin.y (hunk 11 above)
Comment (gawkowsk) [reply Src/spin.y:493]: It is not mine - this change is merged here from nimble repo before I've started any of my changes

File: Src/sched.c (hunk 8 above)
Comment (gawkowsk) [reply Src/sched.c:22]: Agree - but, it is not mine - this change is merged here from nimble repo before I've started any of my changes - nevertheless, it seems like nimble still work around it, and I think changing that would make further merging of their master harder to handle - anyway - your choice...

File: Src/spinlex.c (hunk 14 above)
Comment (gawkowsk) [reply Src/spinlex.c:1735]: nimble called this particular commit as "inline ltl operation abbreviations" ...

File: Src/mesg.c (hunk 2 above)
Comment (gawkowsk) [reply Src/mesg.c:797]: I know - TODO mark was just for me to easy further tracking of potential points in code responsible for queues

File: Src/pangen2.c (hunk 3 above)
Comment (gawkowsk) [reply Src/pangen2.c:1951]: as I've already wrote on Teams - I've just marked for myself points that should be further verified if the value can be only int or maybe float should also be expected. As I will go with queues implementation, the semantics of these points will hopefully be understood.

File: Src/pangen3.c (hunk 5 above)
Comment (gawkowsk) [reply Src/pangen3.c:302]: ok - the comment was left as TODO, why it should be descriptive, why I needed two versions of sr_mesg function...

Comment (gawkowsk) [reply Src/pangen3.c:302]: will be fixed

File: Src/pangen7.c (hunk 6 above)
Comment (gawkowsk) [reply Src/pangen7.c]: so far, in all these points, it seems, that float is not an option - as the queues will be implemented, hopefully TODO will be changed to informative info why I do not border with float const here

File: Src/pangen7.c (hunk 7 above)
Comment (gawkowsk) [reply Src/pangen7.c]: ok

File: Src/spin.h (hunk 9 above)
Comment (gawkowsk) [reply Src/spin.h]: ok

File: Src/spin.h (hunk 10 above)
Comment (gawkowsk) [reply Src/spin.h]: ok

File: Src/spin.h [hunk 23]
```diff
@@ -316,8 +317,9 @@ int	eval(Lextok *);
...
+Value getval(Lextok *);	// TODO PG - actually it takes the value from symbol associated by the token... should be renamed
+float getFloatTokenValue(const Lextok *);
```
Comment (gawkowsk) [inline Src/spin.h]: indent

File: Src/spinlex.c (hunk 13 above)
Comment (gawkowsk) [reply Src/spinlex.c:1614]: actually - we are adapting... we are building more sophisticated finite state automata code to resolve int and float consts... I agree that it should be in the separate function... as almost everything in the lex function... but... they use goto statement inside of the macro which creates the final token... I would personally kill any student who will do something like that... but they did it! moreover, the comment is handled this way which makes me crazy even more... So, finally, after several trials, I gave up and wrote it in their convention.

File: Src/value.h (hunk 16 above)
Comment (gawkowsk) [reply Src/value.h:13]: Yep - historically, this change was done to make a default initialization of values in tokens as int  - otherwise, in some cases I was getting floats by default (e.g. if we have var declared in promela without assignment, spin creates a token node in the ini attribute of the tree with zero value...). I think, this change is not necessary anymore, but I forgot about it as the tests are going fine right now

File: Test/Float/FloatArithmetic.pml (hunk 17 above)
Comment (gawkowsk) [reply Test/Float/FloatArithmetic.pml:22]: If you wish...

File: Test/Float/FloatArithmeticOperators.pml (hunk 18 above)
Comment (gawkowsk) [reply Test/Float/FloatArithmeticOperators.pml]: Yep - will be deleted

File: Test/Float/FloatArithmeticOperators.pml (hunk 19 above)
Comment (gawkowsk) [reply Test/Float/FloatArithmeticOperators.pml]: of course - the reason for 14.0 was the problem with internal casting of float values into int values inside of eval function (run.c) - a result of it was that assert(14.4 == 14) was ok... also assert(14.4 == 14.3) because both values were casted to 14 before comparison... - anyway - I will change it to 14.3 - no problemo

File: Test/Float/FloatArithmeticOperators.pml (hunk 20 above)
Comment (gawkowsk) [reply Test/Float/FloatArithmeticOperators.pml]: not sure if I understand your goal - would you give more example on the idea of your way of printfs?

File: Test/Makefile (hunk 22 above)
Comment (gawkowsk) [reply Test/Makefile]: I think that your comment is actually about the same as "... printf informed about both the values, and their source..." - as you give me more clue on that I would be happy to fix it your way.

**Review by gawkowsk — COMMENTED** [review]
Thank you for the review - I have commented some of yours remarks - please look at them and give ma feedback.

File: .vscode/launch.json (hunk 1 above)
Comment (gawkowsk) [reply .vscode/launch.json]: done - whole .vscode directory is ignored and deleted from the repo

File: Src/pangen7.c (hunk 6 above)
Comment (gawkowsk) [reply Src/pangen7.c]: indentetion fixed

File: Src/pangen7.c (hunk 7 above)
Comment (gawkowsk) [reply Src/pangen7.c]: done

File: Src/spin.h (hunk 9 above)
Comment (gawkowsk) [reply Src/spin.h]: Do we have any doc for our (N7) changes that I should edit? or start a new changelog?

File: Src/spin.h (hunk 10 above)
Comment (gawkowsk) [reply Src/spin.h]: done

File: Src/spin.h (hunk 23 above)
Comment (gawkowsk) [reply Src/spin.h]: done

File: Src/pangen2.c (hunk 4 above)
Comment (gawkowsk) [reply Src/pangen2.c]: done

File: Test/Float/FloatLiteralValues.pml (hunk 21 above)
Comment (gawkowsk) [reply Test/Float/FloatLiteralValues.pml]: hopefully done

File: Test/Float/FloatArithmeticOperators.pml (hunk 20 above)
Comment (gawkowsk) [reply Test/Float/FloatArithmeticOperators.pml]: hopefully done

File: Test/Float/FloatArithmeticOperators.pml (hunk 19 above)
Comment (gawkowsk) [reply Test/Float/FloatArithmeticOperators.pml]: done

File: Test/Float/FloatArithmeticOperators.pml (hunk 18 above)
Comment (gawkowsk) [reply Test/Float/FloatArithmeticOperators.pml]: done

File: Test/Makefile (hunk 22 above)
Comment (gawkowsk) [reply Test/Makefile]: ok - hopefully done

File: Src/sched.c (hunk 8 above)
Comment (Lurkerpas) [reply Src/sched.c:22]: Understood.

File: Src/spin.h (hunk 9 above)
Comment (Lurkerpas) [reply Src/spin.h]: Doc in the code is still not added.
As for changelog - I think that only a single changelog should be maintaned. Changes may be reported once they are fully implemented.

File: Src/spinlex.c (hunk 13 above)
Comment (Lurkerpas) [reply Src/spinlex.c:1614]: Understood....

File: Src/value.h (hunk 16 above)
Comment (Lurkerpas) [reply Src/value.h:13]: In this case, I would remove this change, at least for now - it might be usefull to discover issues with initialization. If at the end we decide that these issues cannot be resolved by the book, we will reconsider adding it.

File: Test/Float/FloatArithmetic.pml (hunk 17 above)
Comment (Lurkerpas) [reply Test/Float/FloatArithmetic.pml:22]: Yes, please. They are a workaround that should no longer be needed.

**Review by Lurkerpas — APPROVED** [review]
